pre_commit/git.py

```diff
@@ -69,7 +74,28 @@ def get_staged_files():
 
 @memoize_by_cwd
 def get_all_files():
-    return cmd_output('git', 'ls-files')[1].splitlines()
+    """Return a list of all actual files in the git repository.
+
+    There are some types of content we want to exclude.  In order to exclude
+    submodules, which git tracks similarly to files, we call `git ls-files
+    --stage` and grep out entries with the special submodule file mode.
+
+    http://stackoverflow.com/a/24122304
+    """
+    # The output format of the command is:
+    # [file mode] [object hash] [stage number]\t[file path]
+    split_regex = re.compile('^([0-7]{6}) [0-9a-f]{40} [0-9]+\t(.+)$')
+
+    def split(line):
+        match = split_regex.match(line)
+        return int(match.group(1), 8), match.group(2)
+
+    output = cmd_output('git', 'ls-files', '--stage')[1]
+    return [
+        path for mode, path in [
+            split(line) for line in output.splitlines()
+        ] if mode != GIT_MODE_SUBMODULE
```

Comment:
Could also exclude symlinks here if desired. I didn't because I can imagine some useful hook (e.g. checking for symlinks pointing outside of the repo), but it could also potentially be dangerous (e.g. we could end up running a hook on a file outside of the git repo, transparently to the hook, which then "fixes" a file outside of the git repo).